src/rvsdg/to_cfg.rs

```diff
@@ -27,54 +27,31 @@ use crate::{
 
 use super::{BasicExpr, Id, Operand, RvsdgBody, RvsdgFunction, RvsdgProgram, RvsdgType};
 
-enum IncompleteBranch {
-    /// Conditional jump from a given block to the next one
-    Cond {
-        from: NodeIndex,
-        var: Identifier,
-        val: bool,
-    },
-    /// Direct jump from a given block to the next one
-    Direct { from: NodeIndex },
-}
-
-impl IncompleteBranch {
-    fn complete(&self, graph: &mut StableDiGraph<BasicBlock, Branch>, to: NodeIndex) {
-        match self {
-            IncompleteBranch::Cond { from, var, val } => {
-                graph.add_edge(
-                    *from,
-                    to,
-                    Branch {
-                        op: BranchOp::Cond {
-                            arg: var.clone(),
-                            val: (*val).into(),
-                        },
-                        pos: None,
-                    },
-                );
-            }
-            IncompleteBranch::Direct { from } => {
-                graph.add_edge(
-                    *from,
-                    to,
-                    Branch {
-                        op: BranchOp::Jmp,
-                        pos: None,
-                    },
-                );
-            }
-        }
-    }
-}
-
 /// Represents the result of a RVSDG computation
 #[derive(Clone, Debug)]
 enum RvsdgValue {
     StateEdge,
     BrilValue(String, Type),
 }
 
+/// The return type of translating a part of an RVSDG
+/// Translating an RVSDG node always creates new blocks.
+/// These may be redundant or have a single input and output, but these
+/// are removed in an optimization pass.
+#[derive(Clone, Debug)]
+struct TranslationResult {
+    start: NodeIndex,
+    end: NodeIndex,
+    values: Vec<RvsdgValue>,
```

Comment:
I think it'd help to include an intuitive translation of what these fields mean, particularly `values`.

In particular, I think the RVSDGs that we currently generate have lots of variables plumbed through that are "redundant" for a given block; if we start hoisting those inputs out, is our use of `values` in this file still correct?

Reply:
Good idea
I added a comment. The idea is that `values` is a list of bindings for the outputs of the node. For single outputs, this is a vec of length one.
To answer your question, I don't see why that would be a problem. If an input is hoisted out, it's still a valid RVSDG and the node will have one less output.